### ISSUE TITLE
Problems with keycloak (openid-conect)

### ISSUE THREAD

**Issue by angelmaturanat** [Worteks/humhub-auth-oidc#1]
Hi! 

I am currently integrating Keycloak with Humhub with you component but i recieve error 400 for response.

`yii\authclient\InvalidResponseException: Request failed with code: 400, message: <html><head><title>Error</title></head><body>Bad Request</body></html> in /var/www/html/humhub/protected/vendor/yiisoft/yii2-authclient/src/BaseOAuth.php:232
Stack trace:
#0 /var/www/html/humhub/protected/vendor/yiisoft/yii2-authclient/src/BaseOAuth.php(340): yii\authclient\BaseOAuth->sendRequest(Object(yii\httpclient\Request))
#1 /var/www/html/humhub/protected/vendor/worteks/yii2auth-oidc/src/OIDC.php(115): yii\authclient\BaseOAuth->api('https://login.c...', 'GET', Array, Array)
#2 /var/www/html/humhub/protected/vendor/yiisoft/yii2-authclient/src/BaseClient.php(174): worteks\yii\authclient\OIDC->initUserAttributes()
#3 /var/www/html/humhub/protected/humhub/modules/user/controllers/AuthController.php(107): yii\authclient\BaseClient->getUserAttributes()
#4 [internal function]: humhub\modules\user\controllers\AuthController->onAuthSuccess(Object(worteks\humhub\authclient\OIDC))
#5 /var/www/html/humhub/protected/vendor/yiisoft/yii2-authclient/src/AuthAction.php(248): call_user_func(Array, Object(worteks\humhub\authclient\OIDC))
#6 /var/www/html/humhub/protected/humhub/modules/user/authclient/AuthAction.php(43): yii\authclient\AuthAction->authSuccess(Object(worteks\humhub\authclient\OIDC))
#7 /var/www/html/humhub/protected/vendor/yiisoft/yii2-authclient/src/AuthAction.php(417): humhub\modules\user\authclient\AuthAction->authSuccess(Object(worteks\humhub\authclient\OIDC))
#8 /var/www/html/humhub/protected/vendor/yiisoft/yii2-authclient/src/AuthAction.php(226): yii\authclient\AuthAction->authOAuth2(Object(worteks\humhub\authclient\OIDC), Array)
#9 /var/www/html/humhub/protected/humhub/modules/user/authclient/AuthAction.php(35): yii\authclient\AuthAction->auth(Object(worteks\humhub\authclient\OIDC))
#10 /var/www/html/humhub/protected/vendor/yiisoft/yii2-authclient/src/AuthAction.php(210): humhub\modules\user\authclient\AuthAction->auth(Object(worteks\humhub\authclient\OIDC))
#11 [internal function]: yii\authclient\AuthAction->run()
#12 /var/www/html/humhub/protected/vendor/yiisoft/yii2/base/Action.php(94): call_user_func_array(Array, Array)
#13 /var/www/html/humhub/protected/vendor/yiisoft/yii2/base/Controller.php(157): yii\base\Action->runWithParams(Array)
#14 /var/www/html/humhub/protected/vendor/yiisoft/yii2/base/Module.php(528): yii\base\Controller->runAction('external', Array)
#15 /var/www/html/humhub/protected/vendor/yiisoft/yii2/web/Application.php(103): yii\base\Module->runAction('user/auth/exter...', Array)
#16 /var/www/html/humhub/protected/vendor/yiisoft/yii2/base/Application.php(386): yii\web\Application->handleRequest(Object(humhub\components\Request))
#17 /var/www/html/humhub/index.php(25): yii\base\Application->run()
#18 {main}`

Analyzing the php code shown in the exception we realize that it performs two executions of the function "return $ this-> sendRequest ()" in the file "yii2-authclient / src / BaseOAuth.php", The first execution gives state 200 but immediately performs another which results in state 400 
resulting in another error that is "Headers already sent".

My common.php file has following:

>                 'oidc' => [
>                         'class' => 'worteks\humhub\authclient\OIDC',
>                         'domain' => 'https://login.domain.cl/auth/realms/master',
>                         'apiBaseUrl' => '/',
>                         'authUrl' => '/protocol/openid-connect/auth',
>                         'tokenUrl' => '/protocol/openid-connect/token',
>                         'userInfoUrl' => 'https://login.domain.cl/auth/realms/master/protocol/openid-connect/userinfo',
>                         'clientId' => 'my-client-id',
>                         'clientSecret' => '640886ce-5a81-4c1b-a6c9-03e275162e1b',
>                         'defaultTitle' => 'Ingreso SSO',
>                         'cssIcon' => 'fa fa-sign-in'
>                 ],
> 

I'm using:

- php: v7.3
- Humhub: v1.3.19
- Keycloak: v7
- Reverse Proxy behind (With apache2)

Thanks and greetings in advance!

**Comment by soisik** [comment]
Hi,
Please check the documentation of the underlying module https://github.com/Worteks/yii2auth-oidc
I would set it up that way
```
'domain' => 'https://login.domain.cl/auth/realms/master',
'authUrl' => '/protocol/openid-connect/auth',
'tokenUrl' => '/protocol/openid-connect/token',
'apiBaseUrl' => '/protocol/openid-connect',
'userInfoUrl' => 'userinfo',
```

**Comment by angelmaturanat** [comment]
Hi! same problem :( 

Img: [https://ibb.co/j3wmrbz](https://ibb.co/j3wmrbz)

I tried to downgrade php from 7.3 to 7.2 and there was no result. Any idea why i get this error?

regards

**Comment by ccabrerareyes** [comment]
Hi! I have the same problem using keycloak and humhub with the same versions. :(

**Comment by angelmaturanat** [comment]
This maybe help to determinate the problem:

```
2020-01-08 15:09:15 [190.8.110.114][-][4qbg1pvndi99b7hu2oo5o1jfej][error][yii\authclient\InvalidResponseException] yii\authclient\InvalidResponseException: Request failed with code: 400, message: <html><head><title>Error</title></head><body>Bad Request</body></html> in /var/www/html/humhub/protected/vendor/yiisoft/yii2-authclient/src/B$Stack trace:
#0 /var/www/html/humhub/protected/vendor/yiisoft/yii2-authclient/src/BaseOAuth.php(342): yii\authclient\BaseOAuth->sendRequest(Object(yii\httpclient\Request))
#1 /var/www/html/humhub/protected/vendor/worteks/yii2auth-oidc/src/OIDC.php(112): yii\authclient\BaseOAuth->api('userinfo', 'GET', Array, Array)
#2 /var/www/html/humhub/protected/vendor/yiisoft/yii2-authclient/src/BaseClient.php(174): worteks\yii\authclient\OIDC->initUserAttributes()
#3 /var/www/html/humhub/protected/humhub/modules/user/controllers/AuthController.php(107): yii\authclient\BaseClient->getUserAttributes()
#4 [internal function]: humhub\modules\user\controllers\AuthController->onAuthSuccess(Object(worteks\humhub\authclient\OIDC))
#5 /var/www/html/humhub/protected/vendor/yiisoft/yii2-authclient/src/AuthAction.php(248): call_user_func(Array, Object(worteks\humhub\authclient\OIDC))
#6 /var/www/html/humhub/protected/humhub/modules/user/authclient/AuthAction.php(43): yii\authclient\AuthAction->authSuccess(Object(worteks\humhub\authclient\OIDC))
#7 /var/www/html/humhub/protected/vendor/yiisoft/yii2-authclient/src/AuthAction.php(417): humhub\modules\user\authclient\AuthAction->authSuccess(Object(worteks\humhub\authclient\OIDC))
#8 /var/www/html/humhub/protected/vendor/yiisoft/yii2-authclient/src/AuthAction.php(226): yii\authclient\AuthAction->authOAuth2(Object(worteks\humhub\authclient\OIDC), Array)
#9 /var/www/html/humhub/protected/humhub/modules/user/authclient/AuthAction.php(35): yii\authclient\AuthAction->auth(Object(worteks\humhub\authclient\OIDC))
#10 /var/www/html/humhub/protected/vendor/yiisoft/yii2-authclient/src/AuthAction.php(210): humhub\modules\user\authclient\AuthAction->auth(Object(worteks\humhub\authclient\OIDC))
#11 [internal function]: yii\authclient\AuthAction->run()
#12 /var/www/html/humhub/protected/vendor/yiisoft/yii2/base/Action.php(94): call_user_func_array(Array, Array)
#13 /var/www/html/humhub/protected/vendor/yiisoft/yii2/base/Controller.php(157): yii\base\Action->runWithParams(Array)
#14 /var/www/html/humhub/protected/vendor/yiisoft/yii2/base/Module.php(528): yii\base\Controller->runAction('external', Array)
#15 /var/www/html/humhub/protected/vendor/yiisoft/yii2/web/Application.php(103): yii\base\Module->runAction('user/auth/exter...', Array)
#16 /var/www/html/humhub/protected/vendor/yiisoft/yii2/base/Application.php(386): yii\web\Application->handleRequest(Object(humhub\components\Request))
#17 /var/www/html/humhub/index.php(25): yii\base\Application->run()
#18 {main}
```

**Comment by coudot** [comment]
Does anyone has any log on KeyCloak side so we can understand why the request is rejected?

**Comment by soisik** [comment]
Hi,

I could reproduce this issue, and will push a fix in a minute (thanks @maxbes for helping figure it out!) A new release will follow.

Thanks for reporting !

**Comment by soisik** [comment]
Fixed in 0.3 release. Please update and reopen if problem persist.
